src/fu-engine-helper.c

```diff
@@ -187,3 +188,106 @@ fu_engine_update_devices_file(FuEngine *self, GError **error)
 	target = g_build_filename(directory, "devices.json", NULL);
 	return g_file_set_contents(target, data, (gssize)len, error);
 }
+
+static void
+fu_engine_integrity_add_measurement(GHashTable *self, const gchar *id, GBytes *blob)
+{
+	g_autofree gchar *csum = g_compute_checksum_for_bytes(G_CHECKSUM_SHA256, blob);
+	g_hash_table_insert(self, g_strdup(id), g_steal_pointer(&csum));
+}
+
+static void
+fu_engine_integrity_measure_acpi(GHashTable *self)
+{
+	g_autofree gchar *path = fu_path_from_kind(FU_PATH_KIND_ACPI_TABLES);
+	const gchar *tables[] = {"SLIC", "MSDM", "TPM2", NULL};
+
+	for (guint i = 0; tables[i] != NULL; i++) {
+		g_autofree gchar *fn = g_build_filename(path, tables[i], NULL);
+		g_autoptr(GBytes) blob = NULL;
+
+		blob = fu_bytes_get_contents(fn, NULL);
+		if (blob != NULL && g_bytes_get_size(blob) > 0) {
+			g_autofree gchar *id = g_strdup_printf("ACPI:%s", tables[i]);
+			fu_engine_integrity_add_measurement(self, id, blob);
+		}
+	}
+}
+
+static void
+fu_engine_integrity_measure_uefi(GHashTable *self)
+{
+	struct {
+		const gchar *guid;
+		const gchar *name;
+	} keys[] = {{FU_EFIVAR_GUID_EFI_GLOBAL, "BootOrder"},
+		    {FU_EFIVAR_GUID_EFI_GLOBAL, "BootCurrent"},
+		    {FU_EFIVAR_GUID_EFI_GLOBAL, "KEK"},
+		    {FU_EFIVAR_GUID_EFI_GLOBAL, "PK"},
+		    {FU_EFIVAR_GUID_SECURITY_DATABASE, "db"},
+		    {FU_EFIVAR_GUID_SECURITY_DATABASE, "dbx"},
+		    {NULL, NULL}};
+
+	/* important keys */
+	for (guint i = 0; keys[i].guid != NULL; i++) {
+		g_autoptr(GBytes) blob =
+		    fu_efivar_get_data_bytes(keys[i].guid, keys[i].name, NULL, NULL);
+		if (blob != NULL) {
+			g_autofree gchar *id = g_strdup_printf("UEFI:%s", keys[i].name);
+			fu_engine_integrity_add_measurement(self, id, blob);
+		}
+	}
+
+	/* Boot#### */
+	for (guint i = 0; i < 0xFF; i++) {
+		g_autofree gchar *name = g_strdup_printf("Boot%04X", i);
+		g_autoptr(GBytes) blob =
+		    fu_efivar_get_data_bytes(FU_EFIVAR_GUID_EFI_GLOBAL, name, NULL, NULL);
+		if (blob != NULL && g_bytes_get_size(blob) > 0) {
+			g_autofree gchar *id = g_strdup_printf("UEFI:%s", name);
+			fu_engine_integrity_add_measurement(self, id, blob);
+		}
+	}
+}
+
+GHashTable *
+fu_engine_integrity_new(GError **error)
+{
+	g_autoptr(GHashTable) self = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, g_free);
+
+	g_return_val_if_fail(error == NULL || *error == NULL, NULL);
+
+	fu_engine_integrity_measure_uefi(self);
+	fu_engine_integrity_measure_acpi(self);
+
+	/* nothing of use */
+	if (g_hash_table_size(self) == 0) {
+		g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND, "no measurements");
+		return NULL;
+	}
```

Comment:
if you follow my other suggestion this block won't matter since you'll always have a measurement for ACPI entries even if not present.